text/0000-layer-metadata-files.md

```diff
@@ -0,0 +1,176 @@
+# Meta
+[meta]: #meta
+- Name: Disambiguate layer metadata files from Application metadata files
+- Start Date: 2021-03-11
+- Author(s): [samj1912](https://github.com/samj1912)
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: (put "N/A" unless this replaces an existing RFC, then link to that RFC)
+
+# Summary
+[summary]: #summary
+
+This is a proposal to change the location of layer specific TOML metadata files from application specific TOML metadata files like `launch.toml` or `build.toml` in order to disambiguate special metadata files from layer metadata files that can have similar names.
+
+# Motivation
+[motivation]: #motivation
+
+Currently all the metadata files live under the `<layers>` folder. This means that application metadata files like `launch.toml`, `build.toml` or `store.toml` share the same directory as layer metadata files. Since the layers could possibly be named `launch`, `build` or `store` this leads to ambiguity, conflicts and/or restrictions on layer names and their associated metadata files that should not exist. This also prevents us from possibly adding new metadata files in the future without breaking backwards compatibility with buildpacks who may have been creating layers with the same name. This proposal seeks to disambiguate special metadata files from free-form layer metadata files.
+
+
+# What it is
+[what-it-is]: #what-it-is
+
+The proposal is as follows - 
+
+The current structure of the `<layers>` folder is -
+
+```
+`/layers` (configurable with `CNB_LAYERS_DIR` or -layers) 
+├── group.toml
+├── plan.toml
+├── <escaped_buildpack_id> (buildpack arg `$1` )
+│   ├── build.toml
+│   ├── launch.toml
+│   ├── store.toml
+│   ├── <layer>.toml
+│   └── <layer>
+└── config
+    └── metadata.toml
+```
+
+
+The proposal is to change this structure to - 
+
+
+```
+`/output` (configurable with `CNB_OUTPUT_DIR` or -output-dir) 
+├── group.toml
+├── plan.toml
+├── layers
+│   └── <escaped_buildpack_id> (`CNB_BP_LAYERS_DIR` )
+│       ├── <layer>.toml
+│       └── <layer>
+└── config
+    ├── metadata.toml
+    └── <escaped_buildpack_id> (`CNB_BP_CONFIG_DIR`  )
+        ├── build.toml
+        ├── launch.toml
+        └── store.toml
+```
+
+The top level `<output-dir>` would be the mounted volume that contains the output from buildpacks. The `<layers>` sub-directory would contains the layer contributions by each buildpack under an `<escaped-buildpack-id>` directory. This directory would be passed to the buildpack via the environment variable `CNB_BP_LAYERS_DIR`. Similarly any config files would reside in the `<config>`, under a sub-directory for each buildpack. This directory would be passed to the buildpack via the environment variable `CNB_BP_CONFIG_DIR`.
+
```

Comment:
We probably need to elaborate on compatibility a bit here. For example, if my build includes two buildpacks, one that implements the new API that includes this change and another older buildpack. If I am using the new platofrm API the directory structure end up looking like like the following?
```
`/output` (configurable with `CNB_OUTPUT_DIR` or -output-dir) 
├── group.toml
├── plan.toml
├── layers
│   └── new-buildpack (`CNB_BP_LAYERS_DIR` )
│       ├── <layer>.toml
│       └── <layer>
│   └── old-buildpack (buildpack arg `$1` )
│       ├── build.toml
│       ├── launch.toml
│       ├── store.toml
│       ├── <layer>.toml
│       └── <layer>
└── config
    ├── metadata.toml
    └── new-buildpack (`CNB_BP_CONFIG_DIR`  )
        ├── build.toml
        ├── launch.toml
        └── store.toml
```

If I am using the old platform API... this gets weirder. We have specified the locations of files like `metadata.toml`, `group.toml`, and the `<layer>` directories and `<layer>.toml` files in the platform API. We haven't specified the location of `build.toml` `launch.toml` etc. at the platform level so we could basically use alternative 4. e.g. 
```
`/layers` (configurable with `CNB_LAYERS_DIR` or -layers-dir) 
├── group.toml
├── plan.toml
├── new-buildpack (`CNB_BP_LAYERS_DIR`)
│   ├── <layer>.toml
│   └── <layer>
├── old-buildpack (buildpack arg `$1` )
│   ├── build.toml
│   ├── launch.toml
│   ├── store.toml
│   ├── <layer>.toml
│   └── <layer>
└── config
    ├── metadata.toml
    └── new-buildpack (`CNB_BP_CONFIG_DIR`  )
        ├── build.toml
        ├── launch.toml
        └── store.toml
```

Reply:
Though not impossible this seems like a lot of edge cases to handle. I could see the debugging experience in these cases being rather interesting :)